lib/hotelsystem.rb

```diff
@@ -0,0 +1,80 @@
+require_relative "reservation"
+
+module Hotel
+  class HotelSystem
+    attr_reader :rooms, :reservations
+
+    def initialize(reservations = [])
+      @rooms = [*1..20]
+      @reservations = reservations
+    end
+
+    def list_rooms
+      room_list = @rooms.map do |num|
+        "Room ##{num}"
+      end
+      return room_list
+    end
+
+    def reservations_by_date(date)
+      reservations_by_date = reservations.select do |booking|
+        booking.in_date_range?(date)
+      end
+      return reservations_by_date
+    end
+
+    def occupied_rooms(days_reservations)
+      occupied_rooms = days_reservations.map do |reservation|
+        reservation.room
+      end
+
+      return occupied_rooms
+    end
+
+    def occupied_rooms_list(start_date, end_date)
+      rooms_occupied = []
+      [*start_date...end_date].each do |date|
+        if reservations_by_date(date)
+          rooms_occupied << occupied_rooms(reservations_by_date(date))
+        end
+      end
+
+      return rooms_occupied.flatten.uniq
+    end
+
+    def room_available?(start_date, end_date, room)
+      return available_rooms_list(start_date, end_date).include?(room)
+    end
+
+
+    def reserve_room(start_date, end_date, room)
+      if invalid_dates?(start_date, end_date)
+        raise ArgumentError, "Invalid Dates. Start date: #{start_date}. End date: #{end_date}."
+      elsif !room_available?(start_date, end_date, room)
+        raise ArgumentError, "Room not available. Room number: #{room}"
+      else
+        reservation = Hotel::Reservation.new(start_date, end_date, room)
+        @reservations << reservation
+        return reservation
+      end
+    end
+
+    def available_rooms_list(start_date, end_date)
+      until invalid_dates?(start_date, end_date)
+        rooms_occupied = occupied_rooms_list(start_date, end_date)
+        avail_rooms_array = @rooms - rooms_occupied
```

Comment:
`until` is used to make a loop, but the body of this loop only runs once. You might want `unless` here instead.

Reply:
I love the problem solving approach here - you've done a good job of taking a big difficult problem (figure out which rooms are free) and turning it into two smaller problems (figure out which rooms are occupied, then take the inverse of that list).